src/Operations/DTO/GenerativeAiOperation.php

```
@@ -126,10 +137,49 @@ public static function getJsonSchema(): array
                             'description' => 'The current state of the operation.',
                         ],
                     ],
-                    'required' => ['id', 'state'],
+                    'required' => [self::KEY_ID, self::KEY_STATE],
                     'additionalProperties' => false,
                 ],
             ],
         ];
     }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @since n.e.x.t
+     *
+     * @return GenerativeAiOperationArrayShape
+     */
+    public function toArray(): array
+    {
+        $data = [
+            self::KEY_ID => $this->id,
+            self::KEY_STATE => $this->state->value,
+        ];
+
+        if ($this->result !== null) {
+            $data[self::KEY_RESULT] = $this->result->toArray();
+        }
+
+        return $data;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @since n.e.x.t
+     */
+    public static function fromArray(array $array): self
+    {
+        static::validateFromArrayData($array, [self::KEY_ID, self::KEY_STATE]);
+
+        $state = OperationStateEnum::from($array[self::KEY_STATE]);
+        $result = null;
+        if (isset($array[self::KEY_RESULT])) {
+            $result = GenerativeAiResult::fromArray($array[self::KEY_RESULT]);
+        }
```

Could validate here that if `KEY_RESULT` is not set but state is "completed", this is wrong and should throw an exception.

---

Ah yup! Resolved in 8ded94d